lib/ecto_adapters_dynamodb/query.ex

```diff
@@ -90,58 +101,85 @@ defmodule Ecto.Adapters.DynamoDB.Query do
         query = construct_search(index, search, opts)
 
         if op == :in do
-          responses_element = "Responses"
-          response_map = %{responses_element => %{table => []}}
+          response_map = %{@responses_element_key => %{table => []}}
 
           Enum.reduce(hash_values, response_map, fn(hash_value, acc) ->
             # When receiving a list of values to query on, construct a custom query for each of those values to pass into do_fetch_recursive/1.
-            %{"Items" => items} = Kernel.put_in(query, [:expression_attribute_values, :hash_key], hash_value)
+            %{"Items" => items} = put_in(query, [:expression_attribute_values, :hash_key], hash_value)
                                   |> (do_fetch_recursive).()
 
-            Kernel.put_in(acc, [responses_element, table], acc[responses_element][table] ++ items)
+            put_in(acc, [@responses_element_key, table], acc[@responses_element_key][table] ++ items)
           end)
         else
           do_fetch_recursive.(query)
         end
       :scan ->
         maybe_scan(table, search, opts)
     end
+  end
+
+  @spec filter(dynamo_response,  non_indexed_filters, String.t) :: dynamo_response | %{}
+  defp filter(%{"Item" => item} = result, non_indexed_filters, _) do
+    if passes_filter?(item, non_indexed_filters),
+      do: result,
+      else: %{}
+  end
+  defp filter(%{@responses_element_key => items} = result, non_indexed_filters, table) do
+    filtered_results = Enum.filter(items[table], &passes_filter?(&1, non_indexed_filters))
+    put_in(result, [@responses_element_key, table], filtered_results)
+  end
+
 
-    results
+  # @spec passes_filter?(dynamo_response, non_indexed_filters) :: boolean()
+  defp passes_filter?(item, [{ logical_op, [ deeper_clauses, filter ] }]) do
+    case logical_op do
+      :and -> evaluate_filter_expression(item, filter) and passes_filter?(item, deeper_clauses)
+      :or -> evaluate_filter_expression(item, filter) or passes_filter?(item, deeper_clauses)
+    end
   end
+  defp passes_filter?(item, [{ _, [ filter ] }]), do: evaluate_filter_expression(item, filter)
+  defp passes_filter?(item, deeper_clauses), do: passes_filter?(item , [ deeper_clauses ])
```

Comment:
I may be behind on the structures. Why are we matching anything (after the other matches failed) and nesting it in a list?